File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* cursor = head;
        ListNode* prev = nullptr;
        ListNode* temp;
        while (cursor) {
            temp = cursor->next;
            cursor->next = prev;
            prev = cursor;
            cursor = temp;
        }
        return prev;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        ll = 0
        cursor = head
        while cursor:
            ll += 1
            cursor = cursor.next
        
        # short-execute logic: if index is out of bounds
        index = ll - n
        if index < 0 or index > ll:
            return head
        
        if n == ll:
            return head.next

        cursor = head
        for i in range(ll):
            if i == index-1:
                if not cursor.next:
                    return head
                cursor.next = cursor.next.next
                break
            cursor = cursor.next
        return head
                    